x-pack/filebeat/input/cel/input.go

```diff
@@ -0,0 +1,909 @@
+
+	client, err := newClient(ctx, cfg, log)
+	if err != nil {
+		return err
+	}
+
+	limiter := newRateLimiterFromConfig(cfg.Resource)
+
+	patterns, err := regexpsFromConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	prg, err := newProgram(cfg.Program, root, client, limiter, patterns)
+	if err != nil {
+		return err
+	}
+
+	var state map[string]interface{}
+	if cfg.State == nil {
+		state = make(map[string]interface{})
+	} else {
+		state = cfg.State
+	}
+	if cursor != nil {
+		state["cursor"] = cursor
+	}
+	goodCursor := cursor
+	goodURL := cfg.Resource.URL.String()
+	state["url"] = goodURL
+	metrics.resource.Set(goodURL)
+	// On entry, state is expected to be in the shape:
+	//
+	// {
+	//     "url": <resource address>,
+	//     "cursor": { ... },
+	//     ...
+	// }
+	//
+	// The url field must be present and can be an HTTP end-point
+	// or a file path. It is currently the responsibility of the
+	// program to handle removing the scheme from a file url if it
+	// is present. The url may be mutated during execution of the
+	// program but the mutated state will not be persisted between
+	// restarts and the url must be present in the returned value
+	// to ensure that it is available in the next evaluation unless
+	// the program has the resource address hard-coded in or it is
+	// available from the cursor.
+	//
+	// Additional fields may be present at the root of the object
+	// and if the program tolerates it, the cursor value may be
+	// absent. Only the cursor is persisted over restarts, but
+	// all fields in state are retained between iterations of
+	// the processing loop except for the produced events array,
+	// see discussion below.
+	//
+	// If the cursor is present the program should perform and
+	// process requests based on its value. If cursor is not
+	// present the program must have alternative logic to
+	// determine what requests to make.
+	//
+	// In addition to this and the functions and globals available
+	// from mito/lib, a global, useragent, is available to use
+	// in requests.
+	err = periodically(ctx, cfg.Interval, func() error {
+		log.Info("process repeated request")
+		var waitUntil time.Time
+		for {
+			// We have a special-case wait for when we have a zero limit.
+			// x/time/rate allow a burst through even when the limit is zero
+			// so in order to ensure that we don't try until we are out of
+			// purgatory we calculate how long we should wait according to
+			// the retry after for a 429 and rate limit headers if we have
+			// a zero rate quota. See handleResponse below.
+			if wait := time.Until(waitUntil); wait > 0 {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-time.After(wait):
+				}
+			}
+
+			// Process a set of event requests.
+			log.Debugw("request state", "state", state)
+			metrics.executions.Add(1)
+			start := time.Now()
+			state, err = evalWith(ctx, prg, map[string]interface{}{
+				root: state,
+			})
+			log.Debugw("response state", "state", state)
```

Comment:
Would `state` include anything sensitive that should not be logged?

Reply:
I'll add a note to the documentation that when logging with debug that the complete state after evaluation is logged and so this should not be used in production. The situation is only slightly worse than the case for logging with debug and seeing all published events, but worth noting.

Reply:
Another option; I think this would make a good enhancement, but not for now because I'd like to spend some more time thinking about the best way to do it, would be to have either a `config.mask` or `state.mask` array of string field that is used to remove/clobber fields in a copy of the state before logging state.